Item highlighting.

Some refactoring and cleaning applied.

From Infernity by quendel, extracted by sergi4ua.

diff --git a/Source/diablo.cpp b/Source/diablo.cpp
index eb44623..a5301c5 100644
--- a/Source/diablo.cpp
+++ b/Source/diablo.cpp
@@ -16,6 +16,7 @@ BOOL gbRunGame;
 int glMid3Seed[NUMLEVELS];
 BOOL gbRunGameResult;
 int zoomflag; // weak
+bool drawitems;
 BOOL gbProcessPlayers;
 int glEndSeed[NUMLEVELS];
 BOOL gbLoadGame;
@@ -1066,6 +1067,9 @@ void ReleaseKey(int vkey)
 {
 	if (vkey == VK_SNAPSHOT)
 		CaptureScreen();
+
+	if (vkey == 0x12)
+		drawitems = false;
 }
 
 void PressKey(int vkey)
@@ -1233,6 +1237,8 @@ void PressKey(int vkey)
 		if(stextflag) {
 			STextNext();
 		}
+	} else if(vkey == 0x12) {
+		drawitems = true;
 	} else if(vkey == VK_LEFT) {
 		if(automapflag && !talkflag) {
 			AutomapLeft();
diff --git a/Source/diablo.h b/Source/diablo.h
index c26a505..558c90c 100644
--- a/Source/diablo.h
+++ b/Source/diablo.h
@@ -95,6 +95,7 @@ extern BOOL gbRunGame;
 extern int glMid3Seed[NUMLEVELS];
 extern BOOL gbRunGameResult;
 extern int zoomflag; // weak
+extern bool drawitems;
 extern BOOL gbProcessPlayers;
 extern int glEndSeed[NUMLEVELS];
 extern BOOL gbLoadGame;
diff --git a/Source/scrollrt.cpp b/Source/scrollrt.cpp
index 1b912ca..a5b8413 100644
--- a/Source/scrollrt.cpp
+++ b/Source/scrollrt.cpp
@@ -1,6 +1,10 @@
 #include "diablo.h"
 #include "../3rdParty/Storm/Source/storm.h"
 
+#include <string>
+#include <vector>
+#include <map>
+
 DEVILUTION_BEGIN_NAMESPACE
 
 int light_table_index;
@@ -390,6 +394,10 @@ void DrawView(int StartX, int StartY)
 
 	DrawDurIcon();
 
+	if(drawitems) {
+		HighlightItemsNameOnMap();
+	}
+
 	if(chrflag) {
 		DrawChr();
 	} else if(questlog) {
@@ -1358,6 +1366,145 @@ void scrollrt_draw_clipped_dungeon_2(BYTE *pBuff, int sx, int sy, int skipChunks
 	}
 }
 
+void HighlightItemsNameOnMap()
+{
+	class itemLabel
+	{
+	public:
+		int itemID;
+		int row;
+		int col;
+		int x;
+		int y;
+		int width;
+		int height;
+		int magicLevel;
+		std::string text;
+		itemLabel(int x, int y, int width, int height, int row, int col, int itemID, int q2, std::string text):
+			itemID(itemID), row(row), col(col), x(x), y(y), width(width), height(height), magicLevel(q2), text(text) {}
+	};
+
+	static const int XOffset = 330;
+	char textOnGround[256];
+	std::vector<itemLabel> q;
+
+	for (int i = 0; i < numitems; i++) {
+		ItemStruct& item_local = item[itemactive[i]];
+		int row = item_local._ix - plr[myplr].WorldX;
+		int col = item_local._iy - plr[myplr].WorldY;
+		// items on ground name highlighting (Qndel)
+		if (item_local._itype == ITYPE_GOLD) {
+			sprintf(textOnGround, "%i gold", item_local._ivalue);
+		}
+		else {
+			sprintf(textOnGround, "%s", item_local._iIdentified ? item_local._iIName : item_local._iName);
+		}
+
+		int walkStandX = ScrollInfo._sxoff;// +plr[myplr]._pyoff;
+		int walkStandY = ScrollInfo._syoff;// +plr[myplr]._pxoff;
+		if (plr[myplr]._pmode == PM_WALK2 && ScrollInfo._sdir == 4) {
+			walkStandX += 32;
+			walkStandY += 16;
+		} else if (plr[myplr]._pmode == PM_WALK2 && ScrollInfo._sdir == 5) {
+			walkStandY +=32;
+		}
+
+		else if(plr[myplr]._pmode == PM_WALK2 && ScrollInfo._sdir == 6) {
+			walkStandX += -32;
+			walkStandY += 16;
+		}
+
+		int x2 = 32 * (row - col) + (200 * (walkStandX) / 100 >> 1);
+		int y2 = 16 * (row + col) + (200 * (walkStandY) / 100 >> 1) - 16;
+																		 //// / (CanRun(myplr) ? 2 : 1);
+		int labelWidth = CalculateTextWidth(textOnGround);
+		int x = x2 - labelWidth/2; // attempt to center the label above the item
+		int y = y2;
+
+		// add to drawing queue
+		int mag = item_local._iMagical;
+		int ix = item_local._ix;
+		int iy = item_local._iy;
+		const std::string &text = std::string(textOnGround);
+		q.push_back(itemLabel(x, y, labelWidth, 13, ix, iy, itemactive[i], mag, text));
+	}
+
+	const int borderX = 5;
+	//bool highlightItem = false;
+	for (unsigned int item1 = 0; item1 < q.size(); ++item1) {
+		std::map<int, bool> backtrace;
+		bool canShow = false;
+		while (!canShow) {
+			canShow = true;
+			for (unsigned int item2 = 0; item2 < item1; ++item2) {
+				if (abs(q[item2].y - q[item1].y) >= q[item1].height + 2) {
+					continue;
+				}
+				if (q[item2].x >= q[item1].x && q[item2].x - q[item1].x < q[item1].width + borderX) {
+					canShow = false;
+					int newpos = q[item2].x - q[item1].width - borderX;
+					if (backtrace.find(newpos) == backtrace.end()) {
+						q[item1].x = newpos;
+						backtrace[newpos] = true;
+					} else {
+						newpos = q[item2].x + q[item2].width + borderX;
+						q[item1].x = newpos;
+						backtrace[newpos] = true;
+					}
+				} else if (q[item2].x < q[item1].x && q[item1].x - q[item2].x < q[item2].width + borderX) {
+					canShow = false;
+					int newpos = q[item2].x + q[item2].width + borderX;
+					if (backtrace.find(newpos) == backtrace.end()) {
+						q[item1].x = newpos;
+						backtrace[newpos] = true;
+					} else {
+						newpos = q[item2].x - q[item1].width - borderX;
+						q[item1].x = newpos;
+						backtrace[newpos] = true;
+					}
+				}
+			}
+		}
+	}
+
+	for (unsigned int i = 0; i < q.size(); ++i) {
+		itemLabel &t = q[i];
+		int bgcolor = 0;
+		// highlight label if item is under cursor:
+		if (pcursitem == t.itemID) {
+			bgcolor = 134;
+		}
+
+		int drawXOffset = -10;
+		if (invflag || sbookflag)
+			drawXOffset -= 160;
+		if (chrflag || questlog)
+			drawXOffset += 160;
+
+		char color = COL_WHITE;
+		if (t.magicLevel == ITEM_QUALITY_MAGIC) {
+			color = COL_BLUE;
+		} else if (t.magicLevel == ITEM_QUALITY_UNIQUE) {
+			color = COL_GOLD;
+		}
+		int sx = t.x + XOffset + drawXOffset;
+		int sy = t.y + 180;
+
+		int sx2 = t.x + XOffset + drawXOffset + 63;
+		int sy2 = t.y + 342;
+
+		if (sx < 0 || sx > 640 || sy < 0 || sy > 480) {
+			continue;
+		}
+
+		if (sx2 < 0 || sx2 > 640 || sy2 < 0 || sy2 > 480) {
+			continue;
+		}
+		DrawSolidRectangle(sx2, t.width + 1, sy2 - t.height, t.height, bgcolor);
+		PrintGameStr(sx, sy, &t.text[0u], color);
+	}
+}
+
 void scrollrt_draw_clipped_e_flag_2(BYTE *pBuff, int x, int y, int skipChunks, signed int CelSkip, int sx, int sy)
 {
 	int lti_old, cta_old, lpi_old;
diff --git a/Source/scrollrt.h b/Source/scrollrt.h
index baeac29..80fe7e6 100644
--- a/Source/scrollrt.h
+++ b/Source/scrollrt.h
@@ -28,6 +28,7 @@ void DrawClippedObject(int x, int y, int ox, int oy, BOOL pre, int CelSkip, int
 void scrollrt_draw_clipped_e_flag(BYTE *pBuff, int x, int y, int sx, int sy);
 void scrollrt_draw_lower_2(int x, int y, int sx, int sy, int chunks, int skipChunks, int eflag);
 void scrollrt_draw_clipped_dungeon_2(BYTE *pBuff, int sx, int sy, int skipChunks, int CelSkip, int dx, int dy, int eflag);
+void HighlightItemsNameOnMap();
 void scrollrt_draw_clipped_e_flag_2(BYTE *pBuff, int x, int y, int skipChunks, signed int CelSkip, int sx, int sy);
 void scrollrt_draw_upper(int x, int y, int sx, int sy, int chunks, int capChunks, int eflag);
 void scrollrt_draw_dungeon(BYTE *pBuff, int sx, int sy, int capChunks, int CelCap, int dx, int dy, int eflag);
diff --git a/Source/town.cpp b/Source/town.cpp
index 6bb1385..2b56405 100644
--- a/Source/town.cpp
+++ b/Source/town.cpp
@@ -1015,6 +1015,9 @@ void T_DrawView(int StartX, int StartY)
 		DrawDiabloMsg();
 	if (PauseMode && !deathflag)
 		gmenu_draw_pause();
+	if(drawitems) {
+		HighlightItemsNameOnMap();
+	}
 	DrawPlrMsg();
 	gmenu_draw();
 	doom_draw();
diff --git a/SourceS/miniwin/misc.h b/SourceS/miniwin/misc.h
index cf4b054..f794897 100644
--- a/SourceS/miniwin/misc.h
+++ b/SourceS/miniwin/misc.h
@@ -763,6 +763,8 @@ constexpr auto DVL_VK_OEM_4 = 0xDB;
 constexpr auto DVL_VK_OEM_5 = 0xDC;
 constexpr auto DVL_VK_OEM_6 = 0xDD;
 constexpr auto DVL_VK_OEM_7 = 0xDE;
+
+constexpr auto DVL_VK_MENU = 0x12;
 //constexpr auto DVL_VK_OEM_8 = 0xDF;
 //constexpr auto DVL_VK_OEM_102 = 0xE2;
 
diff --git a/SourceX/miniwin/misc_msg.cpp b/SourceX/miniwin/misc_msg.cpp
index f39f53a..25aa29e 100644
--- a/SourceX/miniwin/misc_msg.cpp
+++ b/SourceX/miniwin/misc_msg.cpp
@@ -74,6 +74,12 @@ static int translate_sdl_key(SDL_Keysym key)
 	case SDLK_PRINTSCREEN:
 		return DVL_VK_SNAPSHOT;
 
+	case SDLK_MENU:
+		return DVL_VK_MENU;
+
+	case 0x400000E2:
+		return DVL_VK_MENU;
+
 	default:
 		if (sym >= SDLK_a && sym <= SDLK_z) {
 			return 'A' + (sym - SDLK_a);
